src/metrics/codesmells_metrics.nit

```diff
@@ -0,0 +1,268 @@
+		badConceptionElement.add(new LONGPL)
+		badConceptionElement.add(new FEM)
+		badConceptionElement.add(new LONGMETH)
+	end
+end
+
+
+
+class BadConception
+	#Bool Result
+	var result = false
+
+	#Name
+	fun name: String do 
+		return ""
+	end
+
+	#Description
+	fun desc: String do 
+		return ""
+	end
+
+	#Collection method
+	fun collect(classe : AClassdef)do end
+
+	#Show results in console
+	fun printResult do
+		print "{desc} :  {result}"
+	end
+end
+
+
+class Antipattern
+	super BadConception
+end
+
+
+class CodeSmell
+	super BadConception
+end
+
+
+class LARGC
+	super CodeSmell
+
+	redef fun name do 
+		return "LARGC"
+	end
+	redef fun desc do 
+		return "Large class"
+	end
+
+	redef fun collect(aclasse) do
+
+		#get class definition 
+		var classe = aclasse.mclassdef
+		var numberAttribut = classe.collect_intro_mattributes.length
+		#get the number of methods and subtract the get and set of attibutes (numberAtribut*2)
+		var numberMethode = classe.collect_intro_mattributes.length - (numberAttribut*2)
+		if numberMethode > 20 or numberAttribut > 20 then result = true
+	end
+end
+
+class LONGPL
+	super CodeSmell
+
+	var badmethods = new Array[MMethodDef] 
+
+	redef fun name do 
+		return "LONGPL"
+	end
+	redef fun desc do 
+		return "Long parameter list"
+	end
+
+	redef fun collect(aclasse) do
+		#get class definition 
+		var classe = aclasse.mclassdef
+		for meth in classe.mpropdefs do
+			#check if the property is a method definition
+			if meth isa MMethodDef then
+				#Check if method has a signature
+				if meth.msignature != null then
+					if meth.msignature.mparameters.length >= 4 then 
+						badmethods.add(meth)
+						result = true
+					end
+				end
+			end
+		end
+	end
+
+
+	redef fun printResult do
+		print "{desc} :  {result}"
+		if badmethods.length >= 1 then 
+			print "   Affected method :"
+			for method in badmethods do 
+				print "    {method.name}"
+			end
+		end
+	end
+end
+
+
+class FEM
+	super CodeSmell
+
+	var badmethods = new Array[AIdMethid] 
+
+	redef fun name do 
+		return "FEM"
+	end
+	redef fun desc do 
+		return "Feature envy"
+	end
+
+	redef fun collect(aclasse) do
+
+		#Call the visit class method
+		var visits = callvisiteMethodAnalyse(aclasse)
+
+		for visit in visits do
+			if visit.total_call.length < visit.total_call_self.length then
+				result = true
+				badmethods.add(visit.nclassdef.n_methid.as(AIdMethid))
+			end
+		end
+
+	end
+
+
+	redef fun printResult do
+		print "{desc} :  {result}"
+		if badmethods.length >= 1 then 
+			print "   Affected method :"
+			for method in badmethods do 
+				print "    {method.n_id.text}"
+			end
+		end
+	end
+end
+
+class LONGMETH
+	super CodeSmell
+
+	var badmethods = new Array[AIdMethid] 
+
+	redef fun name do 
+		return "LONGMETH"
+	end
+	redef fun desc do 
+		return "Long method"
+	end
+
+	redef fun collect(aclasse) do
+		var visits = callvisiteMethodAnalyse(aclasse)
+
+		for visit in visits do
+			if visit.lineDetail.length > 30 then
```

Comment:
Hard-coded magic number is hard-coded, could you move it to an attribute should we wish to change it at some point?